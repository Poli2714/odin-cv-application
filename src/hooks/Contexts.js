import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
