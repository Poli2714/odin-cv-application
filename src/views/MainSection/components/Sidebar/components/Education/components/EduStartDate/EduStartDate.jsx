import { useContext } from 'react';
import PropTypes from 'prop-types';

import { InputWrapper } from 'src/components/layout';
import { Label, TextInput } from 'src/components/forms';
import { formatDateInput } from 'src/utils';
import { EducationContext } from 'src/hooks/Contexts';

function EduStartDate({ id }) {
  const { educations, setEducations } = useContext(EducationContext);
  const { eduStartDate } = educations[id];

  const handleChangeEduStartDate = e => {
    const formattedDate = formatDateInput(e.target.value, eduStartDate);

    setEducations({
      ...educations,
      [id]: {
        ...educations[id],
        eduStartDate: formattedDate,
      },
    });
  };

  return (
    <InputWrapper>
      <Label labelName="Start" labelFor="eduStartDate" />
      <TextInput
        name="eduStartDate"
        placeholder="MM / YYYY"
        value={eduStartDate}
        onChange={handleChangeEduStartDate}
      />
    </InputWrapper>
  );
}

EduStartDate.propTypes = {
  id: PropTypes.string,
};

export default EduStartDate;
