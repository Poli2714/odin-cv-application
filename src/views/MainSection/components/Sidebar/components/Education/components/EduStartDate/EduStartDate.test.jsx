import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';

import EduStartDate from './EduStartDate';
import { EducationContext } from 'src/hooks/Contexts';

vi.mock('src/components/layout', () => ({
  InputWrapper: vi.fn(({ children }) => (
    <div data-testid="input-wrapper">{children}</div>
  )),
}));

vi.mock('src/components/forms', () => ({
  Label: vi.fn(() => <label htmlFor="eduStartDate">Mock label</label>),
  TextInput: vi.fn(({ onChange, value }) => (
    <input id="eduStartDate" type="text" onChange={onChange} value={value} />
  )),
}));

function MockParentComponent({ id }) {
  const [educations, setEducations] = useState({
    fakeID: { eduStartDate: '' },
    fakeID2: { eduStartDate: '07 / 2017' },
  });

  return (
    <EducationContext.Provider value={{ educations, setEducations }}>
      <EduStartDate id={id} />
    </EducationContext.Provider>
  );
}

test('renders EduStartDate', () => {
  render(<MockParentComponent id="fakeID" />);

  expect(screen.getByTestId('input-wrapper')).toMatchSnapshot();
  expect(screen.getByLabelText('Mock label')).toHaveValue('');
});

test('renders EduStartDate with different initial value', () => {
  render(<MockParentComponent id="fakeID2" />);

  expect(screen.getByLabelText('Mock label')).toHaveValue('07 / 2017');
});

test('updates eduStartDate value as a user types correct date', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="fakeID" />);
  const input = screen.getByLabelText('Mock label');

  await user.type(input, '012022');
  expect(input).toHaveValue('01 / 2022');

  await user.clear(input);
  expect(input).toHaveValue('');
  await user.type(input, '07');
  expect(input).toHaveValue('07 / ');

  await user.type(input, '19');
  expect(input).toHaveValue('07 / 19');
});

test("doesn't update eduStartDate value if a user types incorrect date", async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="fakeID" />);
  const input = screen.getByLabelText('Mock label');

  await user.type(input, '2');
  expect(input).toHaveValue('');

  await user.type(input, '00');
  expect(input).toHaveValue('0');

  await user.type(input, '918');
  expect(input).toHaveValue('09 / 1');

  await user.type(input, '9989');
  expect(input).toHaveValue('09 / 1998');
});
