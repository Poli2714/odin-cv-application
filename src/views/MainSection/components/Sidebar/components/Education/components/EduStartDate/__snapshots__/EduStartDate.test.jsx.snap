// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders EduStartDate 1`] = `
<div
  data-testid="input-wrapper"
>
  <label
    for="eduStartDate"
  >
    Mock label
  </label>
  <input
    id="eduStartDate"
    type="text"
    value=""
  />
</div>
`;
